// @gravity-ui/uikit/build/esm/components/FilePreview/FilePreview.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from 'react';
import { FileZipper as ArchiveIcon, Code as CodeIcon, FileQuestion as DefaultIcon, Picture as ImageIcon, MusicNote as MusicIcon, LogoAcrobat as PdfIcon, LayoutHeaderCellsLarge as TableIcon, TextAlignLeft as TextIcon, Filmstrip as VideoIcon, } from '@gravity-ui/icons';
import { useActionHandlers, useUniqId } from '../../hooks';
import { useBoolean } from '../../hooks/private';
import { Icon } from '../Icon';
import { Text } from '../Text';
import { useMobile } from '../mobile';
import { block } from '../utils/cn';
import { FilePreviewAction } from './FilePreviewAction';
import { MobileImagePreview } from './MobileImagePreview/MobileImagePreview';
import { getFileType } from './utils';
import './FilePreview.css';
const cn = block('file-preview');
const FILE_ICON = {
    default: DefaultIcon,
    image: ImageIcon,
    video: VideoIcon,
    code: CodeIcon,
    archive: ArchiveIcon,
    audio: MusicIcon,
    music: MusicIcon,
    text: TextIcon,
    pdf: PdfIcon,
    table: TableIcon,
};
export function FilePreview({ className, qa, file, imageSrc, description, onClick, actions, }) {
    const id = useUniqId();
    const [previewSrc, setPreviewSrc] = React.useState(imageSrc);
    const [isPreviewSheetVisible, showPreviewSheet, closePreviewSheet] = useBoolean(false);
    const mobile = useMobile();
    const type = getFileType(file);
    const { onKeyDown } = useActionHandlers(onClick);
    React.useEffect(() => {
        if (imageSrc || type !== 'image')
            return undefined;
        try {
            const createdUrl = URL.createObjectURL(file);
            setPreviewSrc(createdUrl);
            return () => {
                URL.revokeObjectURL(createdUrl);
            };
        }
        catch (error) {
            return undefined;
        }
    }, [file, imageSrc, type]);
    const clickable = Boolean(onClick);
    const withActions = Boolean(actions === null || actions === void 0 ? void 0 : actions.length);
    const isPreviewString = typeof previewSrc === 'string';
    const hideActions = isPreviewString && mobile;
    const handleClick = React.useCallback((e) => {
        if (onClick) {
            onClick(e);
            return;
        }
        if (mobile && isPreviewString) {
            showPreviewSheet();
        }
    }, [isPreviewString, mobile, onClick, showPreviewSheet]);
    return (_jsxs("div", { className: cn(null, className), "data-qa": qa, children: [_jsxs("div", { className: cn('card', { clickable, hoverable: clickable || withActions }), role: clickable ? 'button' : undefined, onKeyDown: clickable ? onKeyDown : undefined, tabIndex: clickable ? 0 : undefined, onClick: clickable ? handleClick : undefined, children: [isPreviewString ? (_jsx("div", { className: cn('image-container'), children: _jsx("img", { className: cn('image'), src: previewSrc, alt: file.name }) })) : (_jsx("div", { className: cn('icon', { type }), children: _jsx(Icon, { className: cn('icon-svg'), data: FILE_ICON[type], size: 20 }) })), _jsx(Text, { className: cn('name'), color: "secondary", ellipsis: true, title: file.name, children: file.name }), Boolean(description) && (_jsx(Text, { className: cn('description'), color: "secondary", ellipsis: true, title: description, children: description }))] }), (actions === null || actions === void 0 ? void 0 : actions.length) ? (_jsx("div", { className: cn('actions', { hide: hideActions }), children: actions.map((action, index) => (_jsx(FilePreviewAction, Object.assign({ id: `${id}-${index}` }, action), `${id}-${index}`))) })) : null, _jsx(MobileImagePreview, { visible: isPreviewSheetVisible, onClose: closePreviewSheet, actions: actions, previewSrc: previewSrc, fileName: file.name })] }));
}
FilePreview.displayName = 'FilePreview';
