// @gravity-ui/uikit/build/cjs/components/lab/NumberInput/NumericArrows/NumericArrows.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NumericArrows = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const icons_1 = require("@gravity-ui/icons");
const Button_1 = require("../../../Button");
const Icon_1 = require("../../../Icon");
const layout_1 = require("../../../layout");
const cn_1 = require("../../../utils/cn");
const i18n_1 = tslib_1.__importDefault(require("../i18n"));
const utils_1 = require("../utils");
require("./NumericArrows.css");
const b = (0, cn_1.block)('numeric-arrows');
function NumericArrows(_a) {
    var { className, size, disabled, onUpClick, onDownClick } = _a, restProps = tslib_1.__rest(_a, ["className", "size", "disabled", "onUpClick", "onDownClick"]);
    const commonBtnProps = {
        size: 's',
        pin: 'brick-brick',
        view: 'flat-secondary',
        disabled,
        tabIndex: -1,
        width: 'max',
        extraProps: {
            'aria-hidden': 'true',
        },
    };
    return ((0, jsx_runtime_1.jsxs)(layout_1.Flex, Object.assign({ direction: "column", className: b({ size }, className), qa: utils_1.CONTROL_BUTTONS_QA }, restProps, { children: [(0, jsx_runtime_1.jsx)(Button_1.Button, Object.assign({ className: b('arrow-btn'), qa: utils_1.INCREMENT_BUTTON_QA }, commonBtnProps, { onClick: onUpClick, extraProps: { 'aria-label': (0, i18n_1.default)('label_increment') }, children: (0, jsx_runtime_1.jsx)(Icon_1.Icon, { data: icons_1.ChevronUp, size: 12 }) })), (0, jsx_runtime_1.jsx)("span", { className: b('separator') }), (0, jsx_runtime_1.jsx)(Button_1.Button, Object.assign({ className: b('arrow-btn'), qa: utils_1.DECREMENT_BUTTON_QA }, commonBtnProps, { onClick: onDownClick, extraProps: { 'aria-label': (0, i18n_1.default)('label_decrement') }, children: (0, jsx_runtime_1.jsx)(Icon_1.Icon, { data: icons_1.ChevronDown, size: 12 }) }))] })));
}
exports.NumericArrows = NumericArrows;
