// @gravity-ui/uikit/build/esm/components/Select/components/SelectControl/SelectControl.js
'use client';
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from 'react';
import { ChevronDown, TriangleExclamation } from '@gravity-ui/icons';
import isEmpty from 'lodash/isEmpty';
import { useUniqId } from '../../../../hooks';
import { Icon } from '../../../Icon';
import { Popover } from '../../../Popover';
import { selectControlBlock, selectControlButtonBlock } from '../../constants';
import i18n from '../../i18n';
import { SelectClear } from '../SelectClear/SelectClear';
import { SelectCounter } from '../SelectCounter/SelectCounter';
import './SelectControl.css';
export const SelectControl = React.forwardRef((props, ref) => {
    const { toggleOpen, clearValue, onKeyDown, renderControl, view, size, pin, selectedOptionsContent, className, qa, label, placeholder, isErrorVisible, errorMessage, open, disabled, value, hasClear, popupId, selectId, activeIndex, renderCounter, hasCounter, title, } = props;
    const showOptionsText = Boolean(selectedOptionsContent);
    const showPlaceholder = Boolean(placeholder && !showOptionsText);
    const hasValue = Array.isArray(value) && !isEmpty(value.filter(Boolean));
    const errorTooltipId = useUniqId();
    const [isDisabledButtonAnimation, setIsDisabledButtonAnimation] = React.useState(false);
    const controlMods = {
        open,
        size,
        pin,
        disabled,
        error: isErrorVisible,
        'has-clear': hasClear,
        'no-active': isDisabledButtonAnimation,
        'has-value': hasValue,
    };
    const buttonMods = {
        open,
        size,
        view,
        pin,
        disabled,
        error: isErrorVisible,
    };
    const handleControlClick = React.useCallback((e) => {
        // Fix for Safari
        // https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#clicking_and_focus
        if (e && e.currentTarget !== document.activeElement && 'focus' in e.currentTarget) {
            e.currentTarget.focus();
        }
        toggleOpen();
    }, [toggleOpen]);
    const disableButtonAnimation = React.useCallback(() => {
        setIsDisabledButtonAnimation(true);
    }, []);
    const enableButtonAnimation = React.useCallback(() => {
        setIsDisabledButtonAnimation(false);
    }, []);
    const handleOnClearIconClick = React.useCallback(() => {
        // return animation on clear click
        setIsDisabledButtonAnimation(false);
        clearValue();
    }, [clearValue]);
    const renderCounterComponent = () => {
        if (!hasCounter) {
            return null;
        }
        const count = value.length;
        const counterComponent = _jsx(SelectCounter, { count: count, size: size, disabled: disabled });
        return renderCounter
            ? renderCounter(counterComponent, { count, size, disabled })
            : counterComponent;
    };
    const renderClearIcon = (args) => {
        const valueIsEmpty = value.length === 0;
        if (!hasClear || valueIsEmpty || disabled) {
            return null;
        }
        return (_jsx(SelectClear, { size: size, onClick: handleOnClearIconClick, onMouseEnter: disableButtonAnimation, onMouseLeave: enableButtonAnimation, renderIcon: args.renderIcon }));
    };
    const triggerProps = {
        id: selectId,
        role: 'combobox',
        'aria-controls': open ? popupId : undefined,
        'aria-haspopup': 'listbox',
        'aria-expanded': open,
        'aria-activedescendant': activeIndex === undefined ? undefined : `${popupId}-item-${activeIndex}`,
        onClick: handleControlClick,
        onKeyDown,
        disabled,
    };
    if (renderControl) {
        return renderControl({
            onKeyDown,
            onClear: clearValue,
            onClick: handleControlClick,
            renderClear: renderClearIcon,
            renderCounter: renderCounterComponent,
            ref,
            open,
            popupId,
            selectId,
            activeIndex,
            disabled,
            triggerProps,
        }, { value });
    }
    return (_jsx(React.Fragment, { children: _jsxs("div", { className: selectControlBlock(controlMods), role: "group", children: [_jsxs("button", Object.assign({ ref: ref, className: selectControlButtonBlock(buttonMods, className), type: "button", "data-qa": qa, title: title, tabIndex: 0 }, triggerProps, { children: [label && _jsx("span", { className: selectControlBlock('label'), children: label }), showPlaceholder && (_jsx("span", { className: selectControlBlock('placeholder'), children: placeholder })), showOptionsText && (_jsx("span", { className: selectControlBlock('option-text'), children: selectedOptionsContent }))] })), renderCounterComponent(), renderClearIcon({}), errorMessage && (_jsx(Popover, { content: errorMessage, tooltipId: errorTooltipId, children: _jsx("button", { "aria-label": i18n('label_show-error-info'), "aria-describedby": errorTooltipId, className: selectControlBlock('error-icon'), children: _jsx(Icon, { data: TriangleExclamation, size: size === 's' ? 12 : 16 }) }) })), _jsx(Icon, { className: selectControlBlock('chevron-icon', { disabled }), data: ChevronDown, "aria-hidden": "true" })] }) }));
});
SelectControl.displayName = 'SelectControl';
