// @gravity-ui/uikit/build/cjs/components/ClipboardButton/ClipboardButton.js
"use strict";
'use client';
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClipboardButton = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const React = tslib_1.__importStar(require("react"));
const ActionTooltip_1 = require("../ActionTooltip");
const Button_1 = require("../Button");
const ClipboardIcon_1 = require("../ClipboardIcon");
const CopyToClipboard_1 = require("../CopyToClipboard");
const cn_1 = require("../utils/cn");
const i18n_1 = tslib_1.__importDefault(require("./i18n"));
require("./ClipboardButton.css");
const b = (0, cn_1.block)('clipboard-button');
const DEFAULT_TIMEOUT = 1200;
const TOOLTIP_ANIMATION = 200;
const ButtonSizeToIconSize = {
    xs: 12,
    s: 16,
    m: 16,
    l: 16,
    xl: 20,
};
const ClipboardButtonComponent = (props) => {
    const { size = 'm', hasTooltip = true, tooltipInitialText = (0, i18n_1.default)('startCopy'), tooltipSuccessText = (0, i18n_1.default)('endCopy'), status, view = 'flat', extraProps = {}, children, iconPosition = 'start', closeDelay, onMouseEnter, onFocus } = props, rest = tslib_1.__rest(props, ["size", "hasTooltip", "tooltipInitialText", "tooltipSuccessText", "status", "view", "extraProps", "children", "iconPosition", "closeDelay", "onMouseEnter", "onFocus"]);
    const buttonIcon = ((0, jsx_runtime_1.jsx)(Button_1.Button.Icon, { className: b('icon'), children: (0, jsx_runtime_1.jsx)(ClipboardIcon_1.ClipboardIcon, { size: ButtonSizeToIconSize[size], status: status }) }));
    return ((0, jsx_runtime_1.jsx)(ActionTooltip_1.ActionTooltip, { title: status === 'success' ? tooltipSuccessText : tooltipInitialText, disabled: !hasTooltip, closeDelay: closeDelay, children: (0, jsx_runtime_1.jsxs)(Button_1.Button, Object.assign({ view: view, size: size, extraProps: Object.assign({ 'aria-label': tooltipInitialText }, extraProps), onMouseEnter: onMouseEnter, onFocus: onFocus }, rest, { children: [iconPosition === 'start' ? buttonIcon : null, children, iconPosition === 'end' ? buttonIcon : null] })) }));
};
function ClipboardButton(props) {
    const { text, timeout = DEFAULT_TIMEOUT, onCopy, options, hasTooltip = true, onMouseEnter, onFocus } = props, buttonProps = tslib_1.__rest(props, ["text", "timeout", "onCopy", "options", "hasTooltip", "onMouseEnter", "onFocus"]);
    const timerIdRef = React.useRef();
    const [tooltipCloseDelay, setTooltipCloseDelay] = React.useState(undefined);
    const [tooltipDisabled, setTooltipDisabled] = React.useState(false);
    React.useEffect(() => window.clearTimeout(timerIdRef.current), []);
    const handleCopy = React.useCallback((text, result) => {
        onCopy === null || onCopy === void 0 ? void 0 : onCopy(text, result);
        setTooltipDisabled(false);
        setTooltipCloseDelay(timeout);
        window.clearTimeout(timerIdRef.current);
        timerIdRef.current = window.setTimeout(() => {
            setTooltipDisabled(true);
        }, timeout - TOOLTIP_ANIMATION);
    }, [onCopy, timeout]);
    const resetTooltip = React.useCallback(() => {
        if (tooltipDisabled) {
            setTooltipDisabled(false);
            setTooltipCloseDelay(undefined);
        }
    }, [tooltipDisabled]);
    const handleMouseEnter = React.useCallback((event) => {
        onMouseEnter === null || onMouseEnter === void 0 ? void 0 : onMouseEnter(event);
        resetTooltip();
    }, [onMouseEnter, resetTooltip]);
    const handleFocus = React.useCallback((event) => {
        onFocus === null || onFocus === void 0 ? void 0 : onFocus(event);
        resetTooltip();
    }, [onFocus, resetTooltip]);
    return ((0, jsx_runtime_1.jsx)(CopyToClipboard_1.CopyToClipboard, { text: text, timeout: timeout, onCopy: handleCopy, options: options, children: (status) => ((0, jsx_runtime_1.jsx)(ClipboardButtonComponent, Object.assign({}, buttonProps, { closeDelay: tooltipCloseDelay, hasTooltip: hasTooltip && !tooltipDisabled, status: status, onMouseEnter: handleMouseEnter, onFocus: handleFocus }))) }));
}
exports.ClipboardButton = ClipboardButton;
