// @gravity-ui/uikit/build/cjs/components/Dialog/DialogFooter/DialogFooter.js
"use strict";
'use client';
Object.defineProperty(exports, "__esModule", { value: true });
exports.DialogFooter = void 0;
const tslib_1 = require("tslib");
const jsx_runtime_1 = require("react/jsx-runtime");
const React = tslib_1.__importStar(require("react"));
const Button_1 = require("../../Button");
const Popup_1 = require("../../Popup");
const cn_1 = require("../../utils/cn");
require("./DialogFooter.css");
const b = (0, cn_1.block)('dialog-footer');
function getButtonView(preset) {
    switch (preset) {
        case 'default':
            return 'action';
        case 'success':
            return 'action';
        case 'danger':
            return 'action';
        default:
            return 'action';
    }
}
class DialogFooter extends React.Component {
    constructor() {
        super(...arguments);
        this.errorTooltipRef = React.createRef();
        this.handleKeyDown = (event) => {
            var _a, _b;
            if (event.key === 'Enter') {
                event.preventDefault();
                if (this.props.onClickButtonApply &&
                    !((_a = this.props.propsButtonApply) === null || _a === void 0 ? void 0 : _a.disabled) &&
                    (!this.props.loading || !((_b = this.props.propsButtonApply) === null || _b === void 0 ? void 0 : _b.loading))) {
                    this.props.onClickButtonApply(event);
                }
            }
        };
    }
    componentDidMount() {
        if (this.props.listenKeyEnter) {
            this.attachKeyDownListeners();
        }
    }
    componentDidUpdate(prevProps) {
        if (!this.props.listenKeyEnter && prevProps.listenKeyEnter) {
            this.detachKeyDownListeners();
        }
        if (this.props.listenKeyEnter && !prevProps.listenKeyEnter) {
            this.attachKeyDownListeners();
        }
    }
    componentWillUnmount() {
        this.detachKeyDownListeners();
    }
    render() {
        const { onClickButtonCancel, onClickButtonApply, loading, textButtonCancel, textButtonApply, propsButtonCancel, propsButtonApply, preset, children, errorText, showError, renderButtons, className, } = this.props;
        const buttonCancel = ((0, jsx_runtime_1.jsx)("div", { className: b('button', { action: 'cancel' }), children: (0, jsx_runtime_1.jsx)(Button_1.Button, Object.assign({ view: textButtonApply ? 'flat' : 'normal', size: "l", width: "max", onClick: onClickButtonCancel, disabled: loading }, propsButtonCancel, { children: textButtonCancel })) }));
        const buttonApply = ((0, jsx_runtime_1.jsxs)("div", { className: b('button', { action: 'apply' }), children: [(0, jsx_runtime_1.jsx)(Button_1.Button, Object.assign({ ref: this.errorTooltipRef, type: "submit", view: getButtonView(preset), size: "l", width: "max", onClick: onClickButtonApply, loading: loading, className: b('button-apply', { preset }) }, propsButtonApply, { children: textButtonApply })), errorText && ((0, jsx_runtime_1.jsx)(Popup_1.Popup, { open: showError, anchorRef: this.errorTooltipRef, placement: ['bottom', 'top'], disableLayer: true, disablePortal: true, hasArrow: true, children: (0, jsx_runtime_1.jsx)("div", { className: b('error'), children: errorText }) }))] }));
        return ((0, jsx_runtime_1.jsxs)("div", { className: b(null, className), children: [(0, jsx_runtime_1.jsx)("div", { className: b('children'), children: children }), (0, jsx_runtime_1.jsx)("div", { className: b('bts-wrapper'), children: renderButtons ? (renderButtons(buttonApply, buttonCancel)) : ((0, jsx_runtime_1.jsxs)(React.Fragment, { children: [textButtonCancel && buttonCancel, textButtonApply && buttonApply] })) })] }));
    }
    attachKeyDownListeners() {
        setTimeout(() => {
            window.addEventListener('keydown', this.handleKeyDown);
        }, 0);
    }
    detachKeyDownListeners() {
        window.removeEventListener('keydown', this.handleKeyDown);
    }
}
exports.DialogFooter = DialogFooter;
DialogFooter.defaultProps = {
    preset: 'default',
    showError: false,
    listenKeyEnter: false,
};
